### DISCUSSION TITLE
A proper multi-platform tutorial for CMake

### DISCUSSION THREAD

**Discussion by Lohkdesgds** [liballeg/allegro5#1316] [Ideas]
I've been trying to do Cmake stuff for a long while now. Cmake is the way to easily compile for multiple platforms, right? A well done Cmakelists should be like plug and play on many platforms.

Allegro already has some Cmake stuff going on, but there's too little tutorials about how to do stuff (it is on TO DO [here](https://github.com/liballeg/allegro_wiki/wiki/Using-Visual-Studio-Code#vscode-with-cmake) actually, "actually easier", but how exactly?). I got it compiling the basic stuff after 5+ tries and I'm still confused how to add the addons and so on. I've found an old tutorial [here](http://members.allegro.cc/EdgarReynaldo/BuildA5.html#allegrogit) and that kinda works, but it was not easy finding it. I don't know if it's up to date enough, what I should do, and so on. Some tests seems to just "FAIL" on build, and again, I'm kinda lost here.

One great thing is that there's some pre-built stuff, like the nuget package for Visual Studio. That's awesome, but platform specific. Visual Studio has project-based "projects" while on Linux we do more folder-based or Cmake projects, something like that. At least I do. I can't just copy and paste on Linux and run, that's why it's good for learning about Allegro, but not long term (if you want multi-platform, of course).

Is there some kind of step by step or something to get Allegro working with Cmake? I'm thinking about a way to just "git clone" and run, or minimal steps, because it would be nice for anyone just adding as submodule and making stuff right away.

Thank you.

**Comment by pedro-w** [comment]
It turns out I wrote the original page you linked to (which I don't remember doing...) so I have added some more info on CMake for you.  Hope it helps! To be honest I absolutely hate CMake, I find it unhelpful and irritatingly inconsistent which means I am probably not the best person to advise

**Reply by pedro-w** [reply]
Just thinking it would useful for someone 'skilled in the art' to produce a FindAllegro.cmake along the lines of FindSDL.cmake which we make use of in our own build system. I don't think there is one already.

**Reply by Taurous** [reply]
That would be fantastic!

You can use that same CMakeLists.txt in that wiki page in Windows if you have MSYS2, MinGW, and install Allegro using pacman on the MinGW terminal.

Though that takes a lot more effort than just downloading Allegro binaries and then telling cmake where you put them.

**Reply by Lohkdesgds** [reply]
I'd be nice if there was a way to build Allegro without pkg config, because Windows is... hmm bad? Still, thanks for adding something there, even though my Ubuntu virtual machine got no luck on finding the pkg somehow, but maybe that's me lol

**Comment by connorjclark** [comment]
I have an example CMake that should be cross-platform, and it also is an example of how to build with emscripten. Hopefully it helps.

https://github.com/connorjclark/allegro-project

**Reply by Lohkdesgds** [reply]
Thank you a lot! Seriously, this is exactly what I needed to start doing stuff on CMake 😊

The only thing is that it doesn't copy the dlls, but it works on Windows using Visual Studio 2022 (if I copy the dll manually before running the executable)

Now I just need to know why you did shared libs on MSVC and why shared is the only option working right now haha

**Reply by connorjclark** [reply]
this should copy them over:

```sh
# This allows outputs to be run where they are by putting their libaries in
# the same folder. 
set (CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR})
set (CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR})
```

> Now I just need to know why you did shared libs on MSVC and why shared is the only option working right now haha

if you find out, let me know!

> Thank you a lot! Seriously, this is exactly what I needed to start doing stuff on CMake 😊

Agreed, it would have saved me a couple days early on TBH, since I knew zero CMake a month ago.

**Reply by Lohkdesgds** [reply]
I tried some stuff, but unfortunately I haven't had the time to really make a change here.
```cmake
# ... after FetchContent_Declare
if (MSVC)
	set(SHARED ON)
else()
	set(SHARED OFF)
endif()
set(WANT_MONOLITH OFF)
set(WANT_TESTS OFF)
set(WANT_EXAMPLES OFF)
set(WANT_DEMO OFF)
FetchContent_MakeAvailable(allegro5)
```
this worked exactly the same as your `FetchContent_GetProperties` and `FetchContent_Populate` afaik. `WANT_MONOLITH` I think it's the SHARED thing for allegro 5. Monolith is "not dynamic". I couldn't get it to work nevertheless (it fails to find allegro `.lib` files somehow).
I also tried including all addons with
```cmake
target_link_libraries(al_example LINK_PUBLIC 
	allegro
	allegro_main
	allegro_image
	allegro_primitives
	allegro_color
	allegro_font
	allegro_ttf
	allegro_audio
	allegro_acodec
	allegro_memfile
	allegro_physfs
	allegro_native_dialog
)
```
And I noticed that `ttf`, `physfs` and `native_dialog` won't work (cannot open 'allegro_ttf.lib' and so on).
Meanwhile I had this to copy the headers too
```cmake
# ...
file(COPY ${allegro5_SOURCE_DIR}/addons/ttf/allegro5/allegro_ttf.h
	DESTINATION ${allegro5_SOURCE_DIR}/include/allegro5
)
file(COPY ${allegro5_SOURCE_DIR}/addons/physfs/allegro5/allegro_physfs.h
	DESTINATION ${allegro5_SOURCE_DIR}/include/allegro5
)
file(COPY ${allegro5_SOURCE_DIR}/addons/native_dialog/allegro5/allegro_native_dialog.h
	DESTINATION ${allegro5_SOURCE_DIR}/include/allegro5
)
# ...
```
but still. Maybe I'm just missing something, maybe reading more about cmake this can easily be fixed or something.
Sorry for being lazy (I think I'm being one right now, idk)

I almost forgot, somehow adding the output library things had no change here. I still had to copy dlls around (maybe VS is being dumb lmao)

**Reply by Lohkdesgds** [reply]
I think Visual Studio messed up with its cache 😐

**Reply by humbertodias** [reply]
> I have an example CMake that should be cross-platform, and it also is an example of how to build with emscripten. Hopefully it helps.
> 
> https://github.com/connorjclark/allegro-project

Thanks! It works perfectly for allegro 5.
@connorjclark 
Do you have a similar cmake build for allegro 4?

**Comment by iopiiopi** [comment]
hmmm...  i'm the opposite.  i'm on linux and i just want to know how to use cmake to install allegro5 and all the addons on linux.  (preferably without all the other stuff for other systems).  it's really confusing. 

i had allegro5 and most the addons installed and working before but i was going crazy trying to get the addons i was missing to work.  i uninstalled everything and have been taking a break, just studying general cmake and shell stuff for a couple days now.  i'm almost scared to try and install again.  lol

**Reply by EdgarReynaldo** [reply]
cmake on linux is the easiest because the real steps are done by installing the addons with your package manager. Once that is done then you use cmake to build the make files for allegro. Use ccmake on linux to get a graphical property editor for allegro's cmake file.

**Comment by iopiiopi** [comment]
k, any particular graphical property editor?  i'm on a old laptop with no
gpu.

**Reply by EdgarReynaldo** [reply]
Just try it. ccmake that is.
